cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -56,6 +58,16 @@
     private final PvcOperator pvcOperations;
     private final DeploymentOperator deploymentOperations;
 
+    private KafkaCluster kafka;
+    private Service service;
+    private Service headlessService;
+    private ConfigMap metricsConfigMap;
+    private StatefulSet statefulSet;
+    private Secret clientsCASecret;
+    private Secret clientsPublicSecret;
+    private Secret brokersClientsSecret;
+    private Secret brokersInternalSecret;
```

Comment:
I don't think you can to this. The `KafkaAssemblyOperator` is stateless, and while the lock prevents it running concurrently on the same cluster, I don't think there's anything preventing it running simultaneously across different clusters.

I think you need to make the chained compose() pass this data via the (mostly currently ignored) lambda arguments. You could just simple java bean (hell, I'm not sure even getters are really necessary) to do that. It could be immutable in the above fields, but would need to be mutable in `diffs` and `scale`. That way you could just pass the one thing though the chain of `compose()`s. Wdyt?

Reply:
If I have got your idea ....

First of all, all the above fields plus `service`, `headlessService` and so on seems to be not needed because they come from `kafka.generateXXXXX` which can be used in the `compose()` chain.
It seems to me that just the `kafka` instance is needed.
The `getKafkaCluster` could return a `Future<KafkaCluster>` in order to have the `kafka` instance available to the first compose for calling `kafka.getName()` and `kafka.getReplicas()`.
The first compose could return a `Future<MyBean>` where `MyBean` should bring the `kafka` instance and the result of the `reconcile` in the first `compose`; in this way the `kafka` instance can be used in the next `compose` and so on.
Did I get your point correctly ?